contrib/babelfishpg_tsql/src/pl_exec-2.c

```diff
@@ -2955,6 +2959,225 @@ exec_stmt_insert_bulk(PLtsql_execstate *estate, PLtsql_stmt_insert_bulk *stmt)
+	int64	max_identity_value = 0;
+	int64	cur_identity_value = 0;
+	int		attnum;
+	int64	reseed_value = 0;
+	Oid		nsp_oid;
+	Oid 	table_oid;
+	Oid		seqid = InvalidOid;
+	bool	is_null;
+	volatile bool cur_value_is_null = true;
+
+	if (dbcc_stmt.db_name)
+		db_name = pstrdup(dbcc_stmt.db_name);
+	else
+		db_name = get_cur_db_name();
+
+	if (!DbidIsValid(get_db_id(db_name)))
+	{
+		ereport(ERROR,
+		(errcode(ERRCODE_UNDEFINED_DATABASE),
+			errmsg("database \"%s\" does not exist", db_name)));
+		pfree(db_name);
+	}
+
+	user = get_user_for_database(db_name);
+	guest_role_name = get_guest_role_name(db_name);
+
+	if(dbcc_stmt.new_reseed_value)
+		reseed_value = pg_strtoint64(dbcc_stmt.new_reseed_value);
+
+	ereport(WARNING, (errmsg("reseed_value: %ld", reseed_value)));	// to remove
+
+	if (dbcc_stmt.schema_name)
+	{
+		schema_name = pstrdup(dbcc_stmt.schema_name);
+		nsp_name = get_physical_schema_name(db_name, schema_name);
+	}
+	else
+	{
+		/* 
+		 * If schema_name is not provided, find default schema for current user
+		 * and get physical schema name
+		 */
+		if (!user)
+		{
+			ereport(ERROR,
+			(errcode(ERRCODE_UNDEFINED_OBJECT),
+				 errmsg("user does not exist")));
+		}
+		else if ((guest_role_name && strcmp(user, guest_role_name) == 0))
+		{
+			nsp_name = pstrdup(get_guest_schema_name(db_name));
+		}
+		else
+		{
+			schema_name = get_authid_user_ext_schema_name((const char *) db_name, user);
+			nsp_name = get_physical_schema_name(db_name, schema_name);
+		}
+	}
+
+	pfree(db_name);
+
+	nsp_oid = get_namespace_oid(nsp_name, false);
+	if(!OidIsValid(nsp_oid))
+	{
+		ereport(ERROR,
+		(errcode(ERRCODE_UNDEFINED_SCHEMA),
+			errmsg("schema \"%s\" does not exist", schema_name)));
+	}
+	pfree(schema_name);
+
+	table_oid = get_relname_relid(dbcc_stmt.table_name, nsp_oid);
+	if(!OidIsValid(table_oid))
+	{
+		ereport(ERROR,
+		(errcode(ERRCODE_UNDEFINED_TABLE),
+			errmsg("relation \"%s\" does not exist", dbcc_stmt.table_name)));
+	}
+
+	/* Permission check */
+	if (pg_class_aclcheck(table_oid, GetUserId(), ACL_SELECT | ACL_USAGE) != ACLCHECK_OK)
+	{
+		aclcheck_error(ACLCHECK_NO_PRIV, OBJECT_TABLE, dbcc_stmt.table_name);
+	}
+
+	rel = RelationIdGetRelation(table_oid);
+	tupdesc = RelationGetDescr(rel);
+
+	/* Find Identity column in table and associated sequence */
+	for (attnum = 0; attnum < tupdesc->natts; attnum++)
+	{
+		Form_pg_attribute attr = TupleDescAttr(tupdesc, attnum);
+
+		if (attr->attidentity)
+		{
+			seqid = getIdentitySequence(table_oid, attnum + 1, false);
+			break;
+		}
+	}
+
+	if (!OidIsValid(seqid))
+		ereport(ERROR,
+		(errcode(ERRCODE_UNDEFINED_COLUMN),
+			errmsg("'%s.%s' does not contain an identity column.",
+				nsp_name, dbcc_stmt.table_name)));	
+
+	pfree(nsp_name);
+
+	/*
+	 * We need to find out the last value of the identity column in the table
+	 * and last value of the identity sequence to compare and decide on the
+	 * action that needs to be taken.
+	 */
+	scan = table_beginscan_catalog(rel, 0, NULL);
```

Comment:
Why are we using catalog scan API to scan an user relation?

Reply:
Since we have to use MAX() to find out the maximum value, I have removed the catalog scan API.